File: src/leetcode/java/21.merge-two-sorted-lists.java
/*
 * [21] Merge Two Sorted Lists
 *
 * https://leetcode.com/problems/merge-two-sorted-lists
 *
 * Easy (38.85%)
 * Total Accepted:    
 * Total Submissions: 
 * Testcase Example:  '[]\n[]'
 *
 * Merge two sorted linked lists and return it as a new list. The new list
 * should be made by splicing together the nodes of the first two lists.
 */
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        ListNode n1 = l1;
        ListNode n2 = l2;
        if (n1 == null && n2 == null) return null;

        ListNode head = null;
        ListNode n3 = head;

        while (n1 != null || n2 != null) {
            int nextVal = 0;
            if (n1 == null) {
                nextVal = n2.val;
                n2 = n2.next;
            } else if (n2 == null) {
                nextVal = n1.val;
                n1 = n1.next;
            } else if (n1.val > n2.val) {
                nextVal = n2.val;
                n2 = n2.next;
            } else {
                nextVal = n1.val;
                n1 = n1.next;
            }
            if (head == null) {
                head = new ListNode(nextVal);
                n3 = head;
            } else {
                n3.next = new ListNode(nextVal);
                n3 = n3.next;
            }
        }

        return head;
    }
}

File: leetcode/javascript/21.merge-two-sorted-lists.js
/*
 * [21] Merge Two Sorted Lists
 *
 * https://leetcode.com/problems/merge-two-sorted-lists
 *
 * algorithms
 * Easy (39.02%)
 * Total Accepted:    242.4K
 * Total Submissions: 621K
 * Testcase Example:  '[]\n[]'
 *
 * Merge two sorted linked lists and return it as a new list. The new list
 * should be made by splicing together the nodes of the first two lists.
 */
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var mergeTwoLists = function(l1, l2) {
  if (l1 === null && l2 === null) return null;

  let head = null;
  let prev = null;
  let n1 = l1, n2 = l2;
  while (n1 !== null || n2 !== null) {
    let node;
    if (n1 === null) {
      node = new ListNode(n2.val);
      n2 = n2.next;
    } else if (n2 === null) {
      node = new ListNode(n1.val);
      n1 = n1.next;
    } else if (n1.val < n2.val) {
      node = new ListNode(n1.val);
      n1 = n1.next;
    } else {
      node = new ListNode(n2.val);
      n2 = n2.next;
    }
    if (head === null) head = node;
    if (prev) prev.next = node;
    prev = node;
  }

  return head;
};
